src/modules/qt/gps_drawing.cpp

```diff
@@ -0,0 +1,647 @@
+
+	//original state
+	p.save();
+
+	//draw legend grid lines + text if enabled
+	if ((pdata->graph_type == gpsg_2d_map_graph || pdata->graph_type == gpsg_crop_center_graph) && mlt_properties_get_int(properties, "show_grid"))
+		draw_legend_grid(filter, frame, p, used_crops);
+
+	int i_now = get_now_gpspoint_index(filter, frame);
+	gps_point_proc gps_now = get_now_weighted_gpspoint(filter, frame);
+	point_2d crt_pt = {-1, -1}, next_pt = {-1, -1};
+
+	QPen pen_solid_color0;
+	pen_solid_color0.setWidth( thickness );
+	pen_solid_color0.setColor( colors[0]);
+	pen_solid_color0.setCapStyle( Qt::RoundCap );
+	last_graph_pen = pen_solid_color0;
+
+	QPen pen_solid_color1;
+	pen_solid_color1.setWidth( thickness );
+	pen_solid_color1.setColor( colors[1]);
+	pen_solid_color1.setCapStyle( Qt::RoundCap );
+
+	QPen pen_thin_color1;
+	//1px dots are too small, 1px line feels just too small and sometimes aliaseses weirdly at an angle
+	//meanwhile, I've added the "Two colors" style if the future points/lines are too small, might reconsider this
+	if (thickness > 2)
+		pen_thin_color1.setWidth( 2 );
+	else 
+		pen_thin_color1.setWidth( 1 );
+	pen_thin_color1.setColor( colors[1]);
+	pen_thin_color1.setCapStyle( Qt::RoundCap );//( Qt::SquareCap ); // I don't remember what issue I had that made me use this, keeping it for now
+
+	QPen pen_gradients;
+	pen_gradients.setWidth( thickness );
+	pen_gradients.setCapStyle( Qt::RoundCap );
+
+	//enhancement TODO: if detected gap in GPS data draw a dashed line (but needs a bunch of extra changes)
+	//pen_dash.setStyle( Qt::DashLine );
+
+	//go through the [start_index..end_index] interval of gps_points_p and draw each pair of valid points
+	for (int i=pdata->ui_crops.start_index; i<pdata->ui_crops.end_index; i++) 
+	{
+		int next_i = get_next_valid_gpspoint_index(filter, i);
+		if (i == next_i || get_crtval_bysrc(filter, i) == GPS_UNINIT || get_crtval_bysrc(filter, next_i) == GPS_UNINIT)
+		{
+			// mlt_log_info(filter, "incomplete pair (i=%d, %d) GPS_UNINIT, skipping drawing", i, next_i);
+			continue;
+		}
+
+		crt_pt = get_gpspoint_to_rect(filter, frame, &pdata->gps_points_p[i], rect, used_crops);
+		next_pt = get_gpspoint_to_rect(filter, frame, &pdata->gps_points_p[next_i], rect, used_crops);
+		
+		//don't draw line at all if it is completely outside the rect
+		if ( !rect_intersects_line(qrect, crt_pt, next_pt) )
+			continue;
+
+		//apply color style
+		if (color_style == gpsg_color_by_solid)
+		{
+			p.setPen( pen_solid_color0 );
+		}
+		else if (color_style == gpsg_color_by_solid_past_future)
+		{
+			if (i <= i_now) 
+			{
+				p.setPen( pen_solid_color0 );
+			}
+			else if (i > i_now) 
+				p.setPen( pen_solid_color1 );
+		}
+		else if (color_style == gpsg_color_by_solid_past && i <= i_now || color_style == gpsg_color_by_solid_future && i > i_now)
+		{
+			p.setPen( pen_solid_color0 );
+		}
+		else if (color_style == gpsg_color_by_solid_past && i > i_now || color_style == gpsg_color_by_solid_future && i <= i_now)
```

Comment:
This expression needs more parentheses: it is unreadable to mix && and || without them.